internal/broker/broker.go

```diff
@@ -2,15 +2,738 @@
+		session.firstSelectedMode = authModeID
+	}
+
+	if err = b.updateSession(sessionID, session); err != nil {
+		return nil, err
+	}
+
+	return uiLayoutInfo, nil
+}
+
+func (b *Broker) generateUILayout(session *sessionInfo, authModeID string) (map[string]string, error) {
+	if _, exists := session.allModes[authModeID]; !exists {
+		return nil, fmt.Errorf("selected authentication mode %q does not exist", authModeID)
+	}
+
+	var uiLayout map[string]string
+	switch authModeID {
+	case "password":
+		uiLayout = map[string]string{
+			"type":  "form",
+			"label": "Gimme your password",
+			"entry": "chars_password",
+		}
+
+	case "newpassword":
+		uiLayout = map[string]string{
+			"type":  "newpassword",
+			"label": "Enter your password",
+			"entry": "chars_password",
+		}
+
+	case "qrcode":
+		response, err := b.auth.oauthCfg.DeviceAuth(context.TODO())
+		if err != nil {
+			slog.Error(fmt.Sprintf("Error getting device auth: %v", err))
+			return nil, errors.New("could not generate QR code layout")
+		}
+		session.authInfo["response"] = response
+
+		uiLayout = map[string]string{
+			"type": "qrcode",
+			"label": fmt.Sprintf(
+				"Scan the QR code or access %q and use the code %q",
+				response.VerificationURI,
+				response.UserCode,
+			),
+			"wait":    "true",
+			"button":  "regenerate QR code",
+			"content": response.VerificationURI,
+		}
+	}
+
+	return uiLayout, nil
+}
 
 // IsAuthenticated evaluates the provided authenticationData and returns the authentication status for the user.
-func (b *Broker) IsAuthenticated(sessionID, authenticationData string) (access, data string, err error) {
-	fmt.Println("IsAuthenticated generic OIDC code")
-	getGroups()
-	return "", "", nil
+func (b *Broker) IsAuthenticated(sessionID, authenticationData string) (string, string, error) {
+	session, err := b.getSession(sessionID)
+	if err != nil {
+		return AuthDenied, "", err
+	}
+
+	var authData map[string]string
+	if authenticationData != "" {
+		if err := json.Unmarshal([]byte(authenticationData), &authData); err != nil {
+			return "", "", errors.New("authentication data is not a valid json value")
+		}
+	}
+
+	// Handles the context that will be assigned for the IsAuthenticated handler
+	b.isAuthenticatedCallsMu.Lock()
+	if _, exists := b.isAuthenticatedCalls[sessionID]; exists {
+		b.isAuthenticatedCallsMu.Unlock()
+		return "", "", fmt.Errorf("IsAuthenticated already running for session %q", sessionID)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	b.isAuthenticatedCalls[sessionID] = isAuthenticatedCtx{ctx: ctx, cancelFunc: cancel}
+	b.isAuthenticatedCallsMu.Unlock()
+
+	// Cleans up the IsAuthenticated context when the call is done.
+	defer func() {
+		b.isAuthenticatedCallsMu.Lock()
+		delete(b.isAuthenticatedCalls, sessionID)
+		b.isAuthenticatedCallsMu.Unlock()
+	}()
+
+	authDone := make(chan struct{})
+	var access, data string
+	go func() {
+		access, data = b.handleIsAuthenticated(ctx, &session, authData)
```

Comment:
I agree. How do you feel about having a separate test function for it? (e.g. `TestMultipleIsAuthenticated`)

Since the flow is very different than the one in the "normal" test.

Reply:
As discussed, this will be a separate card (UDENG-2974) so that we can proceed with the review without "wasting" time.